### PR TITLE
fix: validate subprotocol mqtt

### LINKED ISSUE #640: Rumqttc: Validate response header
# Expected Behavior

Rumqttc should validate that the broker that it connects to has the correct websocket-protocol header set. As per the [spec](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Using_WebSocket_as):
` The WebSocket Subprotocol name selected and returned by the Server MUST be “mqtt”`

# Current Behavior

Currently it ignores it and continues not matter the response.

Should be a very simple fix. [The discarded parameter here](https://github.com/bytebeamio/rumqtt/blob/714f71c9bc3369fa186353201dc2e8197d07594c/rumqttc/src/eventloop.rs#L389) includes the response headers, just needs to validate that it is there, otherwise return a error.

### PR DESCRIPTION
**Pull request by Vilayat-Ali**
Fixes #640 

Added validation to check for response headers for websocket subprotocol for "mqtt" and returns a new error `WebsocketSubprotocol` variant if absent.

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] Formatted with `cargo fmt`
- [ ] Make an entry to `CHANGELOG.md` if it's relevant to the users of the library. If it's not relevant mention why.

### CODE REVIEW

File: mosquitto.conf
Comment (de-sh) [inline mosquitto.conf:1]: Remove this file

File: rumqttc/examples/websocket.rs
Comment (de-sh) [inline rumqttc/examples/websocket.rs:1]: Undo these changes

File: rumqttc/src/eventloop.rs [hunk 1]
```diff
@@ -398,7 +401,14 @@ async fn network_connect(
...
 
-            let (socket, _) = async_tungstenite::tokio::client_async(request, tcp_stream).await?;
+            let (socket, response) =
+                async_tungstenite::tokio::client_async(request, tcp_stream).await?;
+
+            response
+                .headers()
+                .get("Sec-WebSocket-Protocol")
+                .filter(|x| x == "mqtt")
+                .ok_or(Err(ConnectionError::WebsocketSubprotocol));
```
Comment (de-sh) [inline rumqttc/src/eventloop.rs:411]:
```suggestion
                .ok_or(Err(ConnectionError::WebsocketSubprotocol))?;
```

File: rumqttc/src/eventloop.rs [hunk 2]
```diff
@@ -411,13 +421,19 @@ async fn network_connect(
...
                 tcp_stream,
                 Some(connector),
             )
             .await?;
 
+            response
+                .headers()
+                .get("Sec-WebSocket-Protocol")
+                .filter(|x| x == "mqtt")
+                .ok_or(Err(ConnectionError::WebsocketSubprotocol));
```
Comment (de-sh) [inline rumqttc/src/eventloop.rs:435]:
```suggestion
                .ok_or(Err(ConnectionError::WebsocketSubprotocol))?;
```

**Review by de-sh — CHANGES_REQUESTED** [review]

File: rumqttc/src/eventloop.rs (hunk 1 above)
Comment (Vilayat-Ali) [reply rumqttc/src/eventloop.rs]: done

File: rumqttc/src/eventloop.rs (hunk 2 above)
Comment (Vilayat-Ali) [reply rumqttc/src/eventloop.rs]: done

**Comment by de-sh** [comment]
@swanandx any plans on fixing clippy suggestions?

**Comment by swanandx** [comment]
> any plans on fixing clippy suggestions?
>
me must haha, should we open separate PR for it ( would need to sync up this branch after merging it )? or just fix it in this one?

> Test against a broker that supports websocket
>
will do!

File: rumqttc/src/websockets.rs [hunk 3]
```diff
@@ -8,6 +10,33 @@ pub enum UrlError {
     Parse(#[from] http::uri::InvalidUri),
 }
 
+#[derive(Debug, thiserror::Error)]
+pub enum ValidationError {
+    #[error("Websocket response does not contain subprotocol header")]
+    SubprotocolHeaderMissing,
+    #[error("Websocket subprotocol header: {0}")]
+    SubprotocolMqtt(String),
```
Comment (swanandx) [inline rumqttc/src/websockets.rs:18]: can we rename this variant to something that will indicate that we can't find "mqtt" in subprotocols list?

File: rumqttc/src/websockets.rs [hunk 4]
```diff
@@ -8,6 +10,33 @@ pub enum UrlError {
...
+        .get("Sec-WebSocket-Protocol")
+        .ok_or(ValidationError::SubprotocolHeaderMissing)?;
+
+    let sub_protocol = val
+        .to_str()
+        .map_err(|_| ValidationError::SubprotocolHeaderMissing)?;
+
+    if sub_protocol != "mqtt" {
+        return Err(ValidationError::SubprotocolMqtt(sub_protocol.to_owned()));
+    }
```
Comment (swanandx) [inline rumqttc/src/websockets.rs:35]: as per specs [ref](https://http.dev/ws#sec-websocket-protocol):
> This header can appear multiple times within the request, or instead, multiple subprotocols can be listed in a comma-delimited format
> 

here, we aren't covering cases where multiple protocols are listed. e.g.
`Sec-WebSocket-Protocol: something-idk, mqtt`

suggestions: we can split the header at `,` and then check if `mqtt` exists!

File: rumqttc/src/websockets.rs [hunk 5]
```diff
@@ -8,6 +10,33 @@ pub enum UrlError {
...
+    response: Response<Option<Vec<u8>>>,
+) -> Result<(), ValidationError> {
+    let val = response
+        .headers()
+        .get("Sec-WebSocket-Protocol")
+        .ok_or(ValidationError::SubprotocolHeaderMissing)?;
+
+    let sub_protocol = val
+        .to_str()
+        .map_err(|_| ValidationError::SubprotocolHeaderMissing)?;
```
Comment (swanandx) [inline rumqttc/src/websockets.rs:31]: if we can't convert it to string ( i.e. it isn't utf8 ) shouldn't mean it was missing right? we can probably use the other variant which shows it wasn't "mqtt"?

**Review by swanandx — COMMENTED** [review]
Tested with rumqttd when it sends "mqtt" and when doesn't ( in which case it closed the connection as expected ). Though this PR isn't covering all the cases ( explained in review comments ), please address them!

Otherwise, great work with PR :100:

**Comment by de-sh** [comment]
@Vilayat-Ali please go through the changes suggested by @swanandx and the last two commits made to fix the PR along the repo code style. Please note the reorg of code and the naming conventions used.

**Pull request merged**